parquet-hadoop/src/main/java/org/apache/parquet/filter2/dictionarylevel/DictionaryFilter.java

```diff
@@ -487,6 +489,92 @@ public <T extends Comparable<T>> Boolean visit(NotIn<T> notIn) {
     return BLOCK_MIGHT_MATCH;
   }
 
+  @Override
+  public <T extends Comparable<T>> Boolean visit(Contains<T> contains) {
+    T value = contains.getValue();
+
+    // Dictionary only contains non-null values, so we can't use it to filter here
+    if (value == null) {
+      return BLOCK_MIGHT_MATCH;
+    }
+
+    Column<T> filterColumn = contains.getColumn();
+    ColumnChunkMetaData meta = getColumnChunk(filterColumn.getColumnPath());
+
+    if (meta == null) {
+      // the column isn't in this file so all values are null, but the value
+      // must be non-null because of the above check.
+      return BLOCK_CANNOT_MATCH;
+    }
+
+    // if the chunk has non-dictionary pages, don't bother decoding the
+    // dictionary because the row group can't be eliminated.
+    if (hasNonDictionaryPages(meta)) {
+      return BLOCK_MIGHT_MATCH;
+    }
+
+    try {
+      Set<T> dictSet = expandDictionary(meta);
+      if (dictSet != null && dictSet.contains(value)) {
+        return BLOCK_MIGHT_MATCH;
+      } else {
+        return BLOCK_CANNOT_MATCH;
+      }
+    } catch (IOException e) {
+      LOG.warn("Failed to process dictionary for filter evaluation.", e);
+    }
+    return BLOCK_MIGHT_MATCH; // cannot drop the row group based on this dictionary
+  }
+
+  @Override
+  public <T extends Comparable<T>> Boolean visit(DoesNotContain<T> doesNotContain) {
+    T value = doesNotContain.getValue();
+
+    Column<T> filterColumn = doesNotContain.getColumn();
+    ColumnChunkMetaData meta = getColumnChunk(filterColumn.getColumnPath());
+
+    if (value == null) {
+      if (meta == null) {
+        // @Todo ??
```

Comment:
Why the TODO? No metadata means the column is not in the file so every value is null. Therefore returning `BLACK_CANNOT_MATCH` is correct.